internal/session.go

```diff
@@ -0,0 +1,421 @@
+	}
+	return *info, nil
+}
+
+func getSessionInfo(ctx Context) *SessionInfo {
+	info := ctx.Value(sessionInfoContextKey)
+	if info == nil {
+		return nil
+	}
+	return info.(*SessionInfo)
+}
+
+func setSessionInfo(ctx Context, sessionInfo *SessionInfo) Context {
+	return WithValue(ctx, sessionInfoContextKey, sessionInfo)
+}
+
+func createSession(ctx Context, creationTasklist string, retryable bool) (Context, error) {
+	if prevSessionInfo := getSessionInfo(ctx); prevSessionInfo != nil && prevSessionInfo.sessionState == sessionStateOpen {
+		return nil, errFoundExistingOpenSession
+	}
+	sessionID, err := generateSessionID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	tasklistChan := GetSignalChannel(ctx, sessionID) // use sessionID as channel name
+	// Retry is only needed when creating new session and the error returned is NewCustomError(errTooManySessionsMsg)
+	retryPolicy := &RetryPolicy{
+		InitialInterval:    time.Second,
+		BackoffCoefficient: 1.1,
+		MaximumInterval:    time.Second * 10,
+		MaximumAttempts:    10,
+		NonRetriableErrorReasons: []string{
+			"cadenceInternal:Panic",
+			"cadenceInternal:Generic",
+			"cadenceInternal:Timeout START_TO_CLOSE",
+			"cadenceInternal:Timeout HEARTBEAT",
+		},
+	}
+	// Creation activity need an infinite StartToCloseTimeout.
+	scheduleToStartTimeout := time.Second * 10
+	startToCloseTimeout := time.Second*math.MaxInt32 - scheduleToStartTimeout
+	HeartbeatTimeout := time.Second * 20
+	ao := ActivityOptions{
+		TaskList:               creationTasklist,
+		ScheduleToStartTimeout: scheduleToStartTimeout,
+		StartToCloseTimeout:    startToCloseTimeout,
+		HeartbeatTimeout:       HeartbeatTimeout,
+	}
+	if retryable {
+		ao.RetryPolicy = retryPolicy
+	}
+	creationCtx := WithActivityOptions(ctx, ao)
+	creationFuture := ExecuteActivity(creationCtx, sessionCreationActivityName, sessionID)
+
+	var creationErr error
+	var tasklist string
+	s := NewSelector(ctx)
+	s.AddReceive(tasklistChan, func(c Channel, more bool) {
+		c.Receive(ctx, &tasklist)
+	})
+	s.AddFuture(creationFuture, func(f Future) {
+		// activity stoped before signal is received, must be creation timeout.
+		creationErr = f.Get(ctx, nil)
+	})
+	s.Select(ctx)
+
+	if creationErr != nil {
+		return nil, creationErr
+	}
+
+	sessionInfo := &SessionInfo{
+		SessionID:    sessionID,
+		tasklist:     tasklist,
+		sessionState: sessionStateOpen,
+	}
+
+	Go(ctx, func(ctx Context) {
+		if creationFuture.Get(ctx, nil) != nil {
+			sessionInfo.sessionState = sessionStateFailed
+		}
+	})
+
+	return setSessionInfo(ctx, sessionInfo), nil
+}
+
+func generateSessionID(ctx Context) (string, error) {
+	var sessionID string
+	err := SideEffect(ctx, func(ctx Context) interface{} {
+		return uuid.New()
+	}).Get(&sessionID)
+	return sessionID, err
+}
+
+func getCreationTasklist(base string) string {
+	return base + "__internal_session_creation"
+}
+
+func getResourceSpecificTasklist(identity, resourceID string) string {
+	return resourceID + "@" + identity
+}
+
```

Comment:
I would add debug level logging of session creation and other status changes.

Reply:
Done.